#!/usr/bin/env python
from scripts.common import ingest_sample_files
from pathlib import Path

# Global configuration options
configfile: 'config.json'
samples_dna = glob_wildcards(join(workpath, '{samples}_R{}.fastq.gz'))
#samples_rna = config['rnareads']
#cohort      = config['coassembly']

rule all:
    input: 
        # DNA input for metagenome assembly
        expand(
            Path(workpath, "rawQC", "{name}.fastq.info.txt"),
            name=samples
        )
    container: "metawrap"


rule metawrap_qc:
    input: 
        expand(Path(workpath, )))
    container: "metamorph_metawrap"